### PR TITLE
Global

### PR DESCRIPTION
**Pull request by klemens-morgenstern**

### CODE REVIEW

File: include/mrdox/Metadata/Source.hpp [hunk 1]
```diff
@@ -49,10 +49,14 @@ struct Location
...
+struct LocationHasFileNamePredicate
+{
+    bool operator()(const Location& loc) { return loc.Filename.empty(); }
+};
+
 /** Like std::optional<SymbolID>
 */
-using OptionalLocation = Optional<Location,
-    decltype([](const Location& loc) { return loc.Filename.empty(); })>;
+using OptionalLocation = Optional<Location, LocationHasFileNamePredicate>;
```
Comment (vinniefalco) [inline include/mrdox/Metadata/Source.hpp:59]: Why?

File: source/Support/GlobalState.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+// this state gets managed by `main`
+struct GlobalState
+{
+    GlobalState();
+    GlobalState(const GlobalState & ) = delete;
+    ~GlobalState();
+
+    // remember: last gets unloaded first.
+    std::unique_ptr<class Generators> generators;
```
Comment (vinniefalco) [inline source/Support/GlobalState.hpp:29]: TIL `class Generators` forward declaration

File: include/mrdox/Metadata/Source.hpp (hunk 1 above)
Comment (klemens-morgenstern) [reply include/mrdox/Metadata/Source.hpp:59]: Warning about internal linkage on gcc

**Comment by vinniefalco** [comment]
@klemens-morgenstern have a look at my _develop_ branch here: https://github.com/vinniefalco/mrdox/commits/develop

Cherry pick that and adjust as needed and then squash/merge.

File: source/Tool/GeneratorsImpl.hpp [hunk 3]
```diff
@@ -8,12 +8,12 @@
 // Official repository: https://github.com/cppalliance/mrdox
 //
 
-#ifndef MRDOX_TOOL_SUPPORT_GENERATORSIMPL_HPP
-#define MRDOX_TOOL_SUPPORT_GENERATORSIMPL_HPP
+#ifndef MRDOX_TOOL_GENERATORSIMPL_HPP
+#define MRDOX_TOOL_GENERATORSIMPL_HPP
```
Comment (vinniefalco) [inline source/Tool/GeneratorsImpl.hpp:12]: `MRDOX_TOOL_TOOL_GENERATORSIMPL_HPP`

File: source/Tool/ToolWorld.hpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+/** Holds the tool's global variables.
+    A single object of this type is used to manage
+    objects whose lifetime and order of destruction
+    is sensitive.
+*/
+class ToolWorld
+{
+    ToolWorld();
+    ToolWorld(const ToolWorld & ) = delete;
```
Comment (vinniefalco) [inline source/Tool/ToolWorld.hpp:29]: why are you explicitly deleting this?

File: source/Tool/ToolWorld.hpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+#include <vector>
+
+namespace clang {
+namespace mrdox {
+
+
+/** Holds the tool's global variables.
+    A single object of this type is used to manage
+    objects whose lifetime and order of destruction
+    is sensitive.
```
Comment (vinniefalco) [inline source/Tool/ToolWorld.hpp:24]: Why did you remove the blank line I inserted after the brief? This is just wrong. The javadoc has a brief and no description:

image

**Comment by klemens-morgenstern** [comment]
Closing as this is needed by and included in #289.
